File: gobus/transport.go
package gotoc

import (
	"encoding/json"
	"errors"
	"fmt"
)

type CalledHandler interface {
	OnCalled(method string, args []interface{}) (interface{}, error)
}

type Transport interface {
	Call(method string, args []interface{}, res interface{}) (interface{}, error)
	Listen(on CalledHandler)
}

type ctransport struct {
	socket   Socket
	onCalled CalledHandler

	debug bool
}

type rpcReq struct {
	Method string        `json:"method"`
	Params []interface{} `json:"params"`
}

type rpcRes struct {
	Error  interface{} `json:"error"`
	Result interface{} `json:"result"`
}

func NewTransportDefault(port int) Transport {
	s := GetSocket()
	return NewTransport(port, s)
}

func NewTransport(port int, s Socket) Transport {
	t := &ctransport{socket: s}
	t.socket.Listen(port, t)
	return t
}

type onCallReply struct {
	ret []byte
}

func (on *onCallReply) OnReply(r []byte) {
	on.ret = r
}

func (c *ctransport) Call(method string, args []interface{}, res interface{}) (interface{}, error) {

	req := rpcReq{Method: method, Params: args}
	in, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("gotoc: marshal err: %s", err)
	}

	//! NOTE Сейчас транспорт синхронный, поэтому всё просто
	onrep := &onCallReply{}
	c.socket.Send(Port_Bus, in, onrep)

	rep := rpcRes{}
	if res != nil {
		rep.Result = res
	}

	err = json.Unmarshal(onrep.ret, &rep)
	if err != nil {
		return nil, fmt.Errorf("gotoc: unmarshal err: %s", err)
	}

	if rep.Error != nil {
		return nil, errors.New(rep.Error.(string))
	}

	return rep.Result, nil
}

func (c *ctransport) Listen(h CalledHandler) {
	c.onCalled = h
}

func (c *ctransport) OnRequest(in []byte, out *[]byte) {

	fmt.Println("gotoc: request: ", string(in))

	if c.onCalled == nil {
		fmt.Println("gotoc: err: not set onCalled")
		return
	}

	req := rpcReq{}
	err := json.Unmarshal(in, &req)
	if err != nil {
		fmt.Println("gotoc: unmarshal err: ", err.Error())
		return
	}

	res := rpcRes{}
	res.Result, err = c.onCalled.OnCalled(req.Method, req.Params)

	if c.debug {
		fmt.Println("gotoc: called ", req.Method, " result: ", res.Result, ", err: ", err)
	}

	if err != nil {
		res.Error = err.Error()
	}

	*out, err = json.Marshal(res)
	if err != nil {
		fmt.Println("gotoc transport: marshal err: ", err)
	}
}

File: qbus/transport.cpp
#include "transport.h"
#include "log.h"
#include <QJsonObject>
#include <QJsonDocument>
#include <QJsonParseError>
#include "gosocket.h"

namespace InterBus {

static QString method_key("method");
static QString args_key("params");
static QString error_key("error");
static QString result_key("result");

struct TransportHandler: public ISocket::Handler
{
    Transport *t;
    TransportHandler(Transport *p) : t(p) {}
    void onRequest(const QByteArray &in, QByteArray *out) { t->onRequest(in, out); }
};

Transport::Transport(int port, ISocket *s)
    : m_port(port), m_socket(s), m_socketHandler(nullptr), m_onCalled(nullptr)
{
    if (!m_socket) {
        m_socket = GoSocket::instance();
    }
}

Transport::~Transport()
{
    listen(nullptr);
}

ISocket *Transport::socket() const
{
    return m_socket;
}

Transport::Ret Transport::call(const QString &method, const QJsonArray &args)
{
    QJsonObject msg;
    msg.insert(method_key, method);
    msg.insert(args_key, args);

    struct OnReply: public GoSocket::OnReply {
        QJsonDocument json;
        QJsonParseError jsonErr;
        void onReply(const QByteArray &r)  {
            qDebug() << "reply:" << r;
            json = QJsonDocument::fromJson(r, &jsonErr);
        }
    } onReply;

    //! NOTE Сейчас транспорт синхронный, поэтому всё просто
    socket()->send(m_port, QJsonDocument(msg).toJson(), &onReply);

    Ret ret;
    if (onReply.jsonErr.error != QJsonParseError::NoError) {
        ret.errorString = "gochannel: internal error " + onReply.jsonErr.errorString();
    }

    if (!ret.success()) {
        return ret;
    }

    QJsonObject retMsg = onReply.json.object();
    ret.errorString = retMsg.value(error_key).toString();
    if (!ret.success()) {
        return ret;
    }

    ret.result = retMsg.value(result_key);
    return ret;
}

bool Transport::listen(Handler *h)
{
    m_onCalled = h;
    if (m_onCalled) {

        m_socketHandler = new TransportHandler(this);
        return m_socket->listen(m_port, m_socketHandler);

    } else {

        m_socket->listen(m_port, nullptr);
        delete m_socketHandler;
    }
    return true;
}

void Transport::onRequest(const QByteArray &in, QByteArray *out)
{
    if (!m_onCalled) {
        return;
    }

    QJsonParseError jsonErr;
    QJsonDocument json = QJsonDocument::fromJson(in, &jsonErr);
    if (jsonErr.error != QJsonParseError::NoError) {
        return;
    }

    QJsonObject req = json.object();
    QString method = req.value(method_key).toString();
    QJsonArray args = req.value(args_key).toArray();

    Ret ret = m_onCalled->onCalled(method, args);

    QJsonObject retMsg;
    if (ret.success()) {
        retMsg.insert(result_key, ret.result);
    } else {
        retMsg.insert(error_key, ret.errorString);
    }

    *out = QJsonDocument(retMsg).toJson();
}

Transport::Handler::~Handler(){}

}
